tensorflow/python/ops/linalg_ops.py

```diff
@@ -539,12 +538,23 @@ def norm(tensor,
 
   with ops.name_scope(name, 'norm', [tensor]):
     tensor = ops.convert_to_tensor(tensor)
+    rank = len(tensor.get_shape().as_list())
+    axis = tuple(map(lambda i: i if i >= 0 else i + rank, axis))
+
     if ord in ['fro', 'euclidean', 2, 2.0]:
-      # TODO(rmlarsen): Move 2-norm to a separate clause once we support it for
-      # matrices.
-      result = math_ops.sqrt(
-          math_ops.reduce_sum(
-              tensor * math_ops.conj(tensor), axis, keepdims=True))
+      if is_matrix_norm and ord in [2, 2.0]:
+        axes = list(range(rank))
+        perm_before = list(filter(lambda i: i not in axis, axes)) + list(axis)
+        perm_after = list(map(perm_before.index, axes))
+        result = array_ops.transpose(array_ops.expand_dims(
```

Comment:
No need to write this as a on-liner. Break it into the three natural steps (permute, compute matrix 2-norms, permute back) to make this more readable.